workflows: add reusable workflow for self-hosted runner

pinkiebell opened the pull request:
Moves the selection and additional wakeup mechanism of the self-hosted runner into a reusable workflow.
It also removes the blocking behaviour for the whole ci workflow file of `consecutive-workflow-action` and replaces it with a github native concurrency group.

pinkiebell commented:
@AronisAt79 

Is it intended that this PR automatically uses the new workflow file? Shouldn't that be not allowed or is it enabled for organization members only?

AronisAt79 commented:
> @AronisAt79
> 
> Is it intended that this PR automatically uses the new workflow file? Shouldn't that be not allowed or is it enabled for organization members only?

I believe this is to allow forks to reuse our CI tasks on github hosted runners (we are using self hosted runners for long duration (~1h) ones)

pinkiebell commented:
> > @AronisAt79
> > Is it intended that this PR automatically uses the new workflow file? Shouldn't that be not allowed or is it enabled for organization members only?
> 
> I believe this is to allow forks to reuse our CI tasks on github hosted runners (we are using self hosted runners for long duration (~1h) ones)

Ah, ok. So it is actually allowed for any pull request to run on the self-hosted runner. I think that should be restricted to either: 1) pull request coming from the same repo 2) or explicitly allowed pull requests (coming from forks), because anyone can open a PR.

What do you think?

AronisAt79 commented:
> > > @AronisAt79
> > > Is it intended that this PR automatically uses the new workflow file? Shouldn't that be not allowed or is it enabled for organization members only?
> > 
> > 
> > I believe this is to allow forks to reuse our CI tasks on github hosted runners (we are using self hosted runners for long duration (~1h) ones)
> 
> Ah, ok. So it is actually allowed for any pull request to run on the self-hosted runner. I think that should be restricted to either: 1) pull request coming from the same repo 2) or explicitly allowed pull requests (coming from forks), because anyone can open a PR.
> 
> What do you think?

it makes sense. For 2, what would be the "run" vs "norun" condition? Labelling from reviewer maybe?  i would also like to have some more opinions on this: @ChihChengLiang @ed255 @CPerezz

pinkiebell commented:
Seems to work (still have to wait). Basically:
* any pull request coming from a different repo (fork, etc) defaults to the github runner `ubuntu-latest`
  The label `self-hosted-ci` has to be assigned to trigger a test with the self hosted runner. This must be done everytime even if the label is also added, then it has to be removed first and added again.
  
* push events and pull requests from the same repo are always running on the self-hosted runner.

pinkiebell commented:
The only thing that bothers me is that the `consecutiveness` step is waiting for pending workflows that are running on the github runner.

ChihChengLiang commented:
@pinkiebell This repo is configured with the option "Require approval for first-time contributors," so untrusted contributors need approvals to trigger the CI. It requires more trust than the `self-hosted-ci` label approach but less intervention.
Do you think we still need the label approach?

ed255 commented:
> @pinkiebell This repo is configured with the option "Require approval for first-time contributors," so untrusted contributors need approvals to trigger the CI. It requires more trust than the `self-hosted-ci` label approach but less intervention. Do you think we still need the label approach?

I also wanted to bring this up.  I think the current settings work well to limit abuse from anyone opening a PR and wasting resources on the self hosted runner.  I'm afraid of the friction that can be introduced by requiring setting the `self-hosted-ci` label, specially because some of our external regular contributors don't have write permissions on the repository, and I I'm not sure if they are able to set the label themselves.

pinkiebell commented:
I removed the label feature and just consolidated the 'experiment' inside it's own workflow file. This basically DRYs the workflow file. Feel free to either review,merge or close this pull request 🙂

ChihChengLiang commented:
@pinkiebell the DRY part looks great. Can we update the title and description to match the proposed change?

pinkiebell commented:
> @pinkiebell the DRY part looks great. Can we update the title and description to match the proposed change?

done

ed255 reviewed (approved):
LGTM! Nice work on unification of tests workflows :)

ChihChengLiang reviewed (approved):
LGTM

CPerezz commented:
Just rebased the branch so that we can merge this :smile:
